### PR TITLE
More Checkpointing Robustness

### PR DESCRIPTION
**Pull request by fabianlim**
- add universal checkpointing

To verify
1. train until a `ds_native` checkpoint is saved, stop the `torchrun` process.
2. restart `torchrun` with a different `num_proc` while using the *exact same train command*, 
3. observe printouts for creating the universal checkpoint in the `latest` directory. Training will resume with the different `num_proc`

Results on Dolomite full FT
- wrong learning rate was inadvertently used here, but it does prove that the method works.

image

Results on Dolomite QLoRA

image

### CODE REVIEW

**Comment by aldopareja** [comment]
@fabianlim, I'm approving, before merging can you include the set of tests you did with the matching plots?, that way we know what doesn't need any more testing.

**Review by aldopareja — APPROVED** [review]

**Comment by aldopareja** [comment]
After including the tests feel free to merge.

**Comment by aldopareja** [comment]
Fabian, I think we should make it optional to load from latest?. What if you want to load from a previous checkpoint?. Or what if you don't want to load at all and just start fresh?.

**Comment by fabianlim** [comment]
> Fabian, I think we should make it optional to load from latest?. What if you want to load from a previous checkpoint?. Or what if you don't want to load at all and just start fresh?.

then you just start from a fresh `output_dir`

**Comment by aldopareja** [comment]
sounds right then, let's merge this

**Comment by aldopareja** [comment]
the linter is complaining, can you fix that?

**Comment by russellb** [comment]
In the future, can you squash commits like "improvements" and "lint + format" into previous commits prior to merging? Thanks!
